core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala

```diff
@@ -190,4 +191,73 @@ class TSqlExpressionBuilder extends TSqlParserBaseVisitor[ir.Expression] with Pa
     val args = Option(ctx.expression()).map(_.asScala.map(_.accept(this))).getOrElse(Seq.empty)
     FunctionBuilder.buildFunction(name, args)
   }
+
+  // Note that this visitor is made complicated and difficult because the built in ir does not use options
+  // and so we build placeholder values for the optional values. They also do not extend expression
+  // so we can't build them logically with visit and accept. Maybe replace them with
+  // extensions that do do this?
+  override def visitExprOver(ctx: ExprOverContext): ir.Window = {
+    val windowFunction = ctx.expression().accept(this)
+    val partitionByExpressions =
+      Option(ctx.overClause().expression()).map(_.asScala.toList.map(_.accept(this))).getOrElse(List.empty)
+    val orderByExpressions = Option(ctx.overClause().orderByClause())
+      .map(_.orderByExpression().asScala.toList.map { orderByExpr =>
+        val expression = orderByExpr.expression().accept(this)
+        val sortOrder =
+          if (Option(orderByExpr.DESC()).isDefined) ir.DescendingSortDirection
+          else ir.AscendingSortDirection
+        ir.SortOrder(expression, sortOrder, ir.SortNullsUnspecified)
+      })
+      .getOrElse(List.empty)
+
+    val rowRange = Option(ctx.overClause().rowOrRangeClause())
+      .map(buildWindowFrame)
+      .getOrElse(noWindowFrame)
+
+    ir.Window(windowFunction, partitionByExpressions, orderByExpressions, rowRange)
+  }
+
+  private def noWindowFrame: ir.WindowFrame =
+    ir.WindowFrame(
+      ir.UndefinedFrame,
+      ir.FrameBoundary(current_row = false, unbounded = false, ir.Noop),
+      ir.FrameBoundary(current_row = false, unbounded = false, ir.Noop))
+
+  private def buildWindowFrame(ctx: RowOrRangeClauseContext): ir.WindowFrame = {
+    val frameType = buildFrameType(ctx)
+    val fs = Trees.findAllRuleNodes(ctx, TSqlParser.RULE_windowFrameBound)
+
+    fs.headOption match {
+      case Some(frameStartNode: WindowFrameBoundContext) =>
+        val frameStart = buildFrame(frameStartNode)
+        val frameEnd = fs.tail.headOption match {
+          case Some(frameEndNode: WindowFrameBoundContext) => buildFrame(frameEndNode)
+          case _ => ir.FrameBoundary(current_row = false, unbounded = false, ir.Noop)
+        }
+        ir.WindowFrame(frameType, frameStart, frameEnd)
+      case _ => // Cannot reach
+        noWindowFrame
+    }
```

Comment:
This part could be extracted as a separate`private [tsql]` function taking a `FrameType` and a `Collection[ParseTree]` (or some scala counterpart of `Collection[ParseTree]`), making it amenable to unit test covering the unreachable state (using mockito).

```suggestion
    fs.headOption match {
      case Some(frameStartNode: WindowFrameBoundContext) =>
        val frameStart = buildFrame(frameStartNode)
        val frameEnd = fs.tail.headOption match {
          case Some(frameEndNode: WindowFrameBoundContext) => buildFrame(frameEndNode)
          case _ => ir.FrameBoundary(current_row = false, unbounded = false, ir.Noop)
        }
        ir.WindowFrame(frameType, frameStart, frameEnd)
      case _ => // Cannot reach
        noWindowFrame
    }
```

Reply:
Reworked as new function that avoids redundant case altogether